### PR TITLE
FEATURE: Add `ContentSubgraphInterface::findClosestNode()`

### LINKED ISSUE #4517: FEATURE: Add `SubgraphInterface::findClosestNode()`
Currently there is still a lot of code like this:

```php
protected function getClosestDocumentNode(Node $node): ?Node
{
    $subgraph = $this->contentRepositoryRegistry->subgraphForNode($node);
    while ($node !== null && !$node->nodeType->isOfType('Neos.Neos:Document')) {
        $node = $subgraph->findParentNode($node->nodeAggregateId);
    }
    return $node;
}
```

This _could_ be replaced, by
```php
$subgraph->findAncestorNodes($node->nodeAggregateId, FindAncestorNodesFilter::create(nodeTypeConstraints: 'Neos.Neos:Document')->first();
```
but that is a bit cumbersome and not the most performant way (because it potentially loads more results than necessary)

Therefore we should add a dedicated finder that allows for:

```php
$subgraph->findClosestAncestorNode($node->nodeAggregateId, FindClosestAncestorNodeFilter::create(nodeTypeConstraints: 'Neos.Neos:Document');
```

### PR DESCRIPTION
**Pull request by bwaidelich**
Resolves: #4517

### CODE REVIEW

**Comment by mhsdesign** [comment]
Hi, just a though, what about naming it `findClosest` instead and possibly include the current node in the result set if it matches already?

Then this would be an actual 1 to 1 drop in for `findClosestDocumentNode` as we want return the current node if its an document already.

```php
$documentNode = $subgraph->findClosestNode($node->nodeAggregateId, FindClosestNodeFilter::create(nodeTypeConstraints: 'Neos.Neos:Document');
```

**Comment by bwaidelich** [comment]
@mhsdesign that's a very good point, that makes the query much more useful. Adjusted with cddc41dc31e63d3b76ce723b05a0380cc8694532

File: Neos.ContentRepository.Core/Classes/Projection/ContentGraph/ContentSubgraphInterface.php [hunk 1]
```diff
@@ -109,6 +109,14 @@ public function findAncestorNodes(NodeAggregateId $entryNodeAggregateId, Filter\
      */
     public function countAncestorNodes(NodeAggregateId $entryNodeAggregateId, Filter\CountAncestorNodesFilter $filter): int;
 
+    /**
+     * Find the first matching node above the $entryNodeAggregateId that match the specified $filter and return it
+     * Note: in contrast to {@see findAncestorNodes()} the resulting node will be the entry node if it matches the filter!
+     *
+     * @return Node|null the closest ancestor node that matches the given $filter, or NULL if no matching ancestor node was found
```
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Projection/ContentGraph/ContentSubgraphInterface.php]:
```suggestion
     * Find the closest matching node, the entry node itself or the first ancestors of the $entryNodeAggregateId, that match the specified $filter and return it
     * Note: in contrast to {@see findAncestorNodes()} the resulting node will be the entry node if it matches the filter!
     *
     * @return Node|null the closest node that matches the given $filter, or NULL if no matching node was found
```

This was a bit misunderstanding

**Comment by mhsdesign** [comment]
before merging this i would like to discuss how we want to ideally fetch all the necessary nodes for the fusion view.

Based on an entryNode, we need `documentNode` and `siteNode`.

So the document node can be found via this PR `findClosestNode` - but what about the site?

I had the idea to combine both lookups into one query:

```php
$ancestors = $subgraph->findAncestorNodes($node->nodeAggregateId, FindAncestorNodesFilter::create(nodeTypeConstraints: 'Neos.Neos:Document');

$documentNode = $node->nodeType->isOfType('Neos.Neos:Document') ? $node : $ancestors->first();

$siteNode = $ancestors->last();
```

but i dont know if its a criteria for the homepage node to be of type Neos.Neos:Document? see https://neos-project.slack.com/archives/C3MCBK6S2/p1694339628529459?thread_ts=1694269188.257739&channel=C3MCBK6S2&message_ts=1694339628.529459

--------

the other approach using your `findClosestNode` and this hacky `SiteNodeUtility::findSiteNode` 
https://github.com/neos/neos-development-collection/blob/d9c374afb1b73267973e53e3789b949d45cb64df/Neos.Neos/Classes/Domain/Service/SiteNodeUtility.php#L40-L57

```php
$documentNode =  $subgraph->findClosestNode($node->nodeAggregateId, FindClosestNodeFilter::create(nodeTypeConstraints: 'Neos.Neos:Document');

$siteNode = $this->siteNodeUtility->findSiteNode($node);
```

-------

i just remembered that Bernhard hat the plan to introduce a `Neos.Neos:Site` root site nodeType: https://github.com/neos/neos-development-collection/issues/4531

this would allow to write a querys like:

```php
$documentNode =  $subgraph->findClosestNode($node->nodeAggregateId, FindClosestNodeFilter::create(nodeTypeConstraints: 'Neos.Neos:Document');

$siteNode =  $subgraph->findClosestNode($node->nodeAggregateId, FindClosestNodeFilter::create(nodeTypeConstraints: 'Neos.Neos:Site');
```

which is imo as perfect as it gets?

**Comment by bwaidelich** [comment]
> before merging this i would like to discuss how we want to ideally fetch all the necessary nodes for the fusion view

is that really related to this very PR?

> the other approach using your findClosestNode and this hacky SiteNodeUtility::findSiteNode

Why is that hacky?

**Comment by skurfuerst** [comment]
Hey, 

looked at the code, looks great to me.

@bwaidelich IMHO this can be merged. Great work :)

@mhsdesign I do not understand why you want to open up the use cases even more; IMHO these are separate topics.

All the best,
Sebastian

**Review by skurfuerst — APPROVED** [review]

**Review by mhsdesign — APPROVED** [review]

**Comment by mhsdesign** [comment]
yes no sorry all fine - that was just the thinking progress in my head ^^

i really like this and lets merge it!

**Pull request merged**